spacy/tokens/doc.pyx

```diff
@@ -1650,22 +1669,48 @@ cdef class Doc:
             for span_group in self.spans:
                 data["spans"][span_group] = []
                 for span in self.spans[span_group]:
-                    span_data = {
-                        "start": span.start_char, "end": span.end_char, "label": span.label_, "kb_id": span.kb_id_
-                    }
+                    span_data = {"start": span.start_char, "end": span.end_char, "label": span.label_, "kb_id": span.kb_id_}
                     data["spans"][span_group].append(span_data)
 
         if underscore:
+            doc_keys = set()
+            user_keys = set()
+            if self.user_data:
+                data["user_data"] = {"spans":{}, "tokens":{}}
+                for data_key in self.user_data:
+                    if type(data_key) == tuple and data_key[0] == "._.":
+                        _, attr, start, end = data_key
+                        if attr in underscore:
+                            value = self.user_data[data_key]
+                            if not srsly.is_json_serializable(value):
+                                raise ValueError(Errors.E107.format(attr=attr, value=repr(value)))
+                            # Check if doc attribute
+                            if start is None:
+                                doc_keys.add(attr)
+                                continue
+                            # Check if token attribute
+                            elif end is None:
+                                if attr not in data["user_data"]["tokens"]:
+                                    data["user_data"]["tokens"][attr] = {"token_start": start, "value":value}
+                            # Else span attribute
+                            else:
+                                if attr not in data["user_data"]["spans"]:
+                                    data["user_data"]["spans"][attr] = {"token_start": start, "token_end": end, "value":value}
```

Comment:
I would represent both token and span as the same way they are internally with `token_end` as `None` for tokens.